### PR TITLE
Make ipnetwork and ipaddress hashable

### PR DESCRIPTION
**Pull request by Miauwkeru**

### CODE REVIEW

File: flow/record/fieldtypes/net/ip.py [hunk 1]
```diff
@@ -8,49 +11,55 @@ class ipaddress(FieldType):
     val = None
     _type = "net.ipaddress"
 
-    def __init__(self, addr):
+    def __init__(self, addr: Union[str, int]):
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:14]:
```suggestion
    def __init__(self, addr: str | int):
```

File: flow/record/fieldtypes/net/ip.py [hunk 2]
```diff
@@ -1,4 +1,7 @@
+from __future__ import annotations
+
 from ipaddress import ip_address, ip_network
+from typing import Union
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:4]:
```suggestion
```

File: flow/record/fieldtypes/net/ip.py [hunk 3]
```diff
@@ -8,49 +11,55 @@ class ipaddress(FieldType):
     val = None
     _type = "net.ipaddress"
 
-    def __init__(self, addr):
+    def __init__(self, addr: Union[str, int]):
         self.val = ip_address(addr)
 
-    def __eq__(self, b):
+    def __eq__(self, b: Union[str, int]) -> bool:
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:17]:
```suggestion
    def __eq__(self, b: str | int) -> bool:
```

File: flow/record/fieldtypes/net/ip.py [hunk 4]
```diff
@@ -8,49 +11,55 @@ class ipaddress(FieldType):
...
+    def _unpack(data: int) -> ipaddress:
         return ipaddress(data)
 
 
 class ipnetwork(FieldType):
     val = None
     _type = "net.ipnetwork"
 
-    def __init__(self, addr):
+    def __init__(self, addr: Union[str, int]):
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:49]:
```suggestion
    def __init__(self, addr: str | int):
```

File: flow/record/fieldtypes/net/ip.py [hunk 5]
```diff
@@ -8,49 +11,55 @@ class ipaddress(FieldType):
...
         self.val = ip_network(addr)
 
-    def __eq__(self, b):
+    def __eq__(self, b: Union[str, int]) -> bool:
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:52]:
```suggestion
    def __eq__(self, b: str | int) -> bool:
```

File: flow/record/fieldtypes/net/ip.py [hunk 6]
```diff
@@ -59,23 +68,23 @@ def _is_subnet_of(a, b):
         except AttributeError:
             raise TypeError("Unable to test subnet containment " "between {} and {}".format(a, b))
 
-    def __contains__(self, b):
+    def __contains__(self, b: Union[str, int, ip_address]) -> bool:
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:71]:
```suggestion
    def __contains__(self, b: str | int | ip_address) -> bool:
```

**Review by Schamper — DISMISSED** [review]

File: flow/record/fieldtypes/net/ip.py [hunk 7]
```diff
@@ -59,23 +68,23 @@ def _is_subnet_of(a, b):
         except AttributeError:
             raise TypeError("Unable to test subnet containment " "between {} and {}".format(a, b))
```
Comment (Schamper) [inline flow/record/fieldtypes/net/ip.py:69]: Can also f-string this if we're modernizing anyway.

File: tests/test_fieldtype_ip.py [hunk 8]
```diff
@@ -90,6 +90,7 @@ def test_record_ipnetwork():
     assert "192.168.1.1" not in r.subnet
     assert isinstance(r.subnet, net.ipnetwork)
     assert repr(r.subnet) == "net.ipnetwork('192.168.0.0/24')"
+    assert hash(r.subnet) == hash(net.ipnetwork("192.168.0.0/24"))
```
Comment (yunzheng) [inline tests/test_fieldtype_ip.py:93]: can you also add a test with set() ?

**Review by yunzheng — CHANGES_REQUESTED** [review]

File: flow/record/fieldtypes/net/ip.py (hunk 1 above)
Comment (yunzheng) [inline flow/record/fieldtypes/net/ip.py:14]: Actually ipaddress can also initialize from 4 bytes, and can you change it to new style typing? eg: `str | int | bytes`

https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address

**Review by yunzheng — CHANGES_REQUESTED** [review]
can you change to new style typing

**Comment by yunzheng** [comment]
> > can you change to new style typing
> 
> it will break the python3.8 compatibility if I change it to new style typing, which is why i haven't done so.

should work fine due to `from __future__ import annotations`

File: flow/record/fieldtypes/net/ip.py [hunk 9]
```diff
@@ -1,56 +1,75 @@
...
 class ipaddress(FieldType):
     val = None
     _type = "net.ipaddress"
 
-    def __init__(self, addr):
+    def __init__(self, addr: str | int | bytes):
         self.val = ip_address(addr)
 
-    def __eq__(self, b):
+    def __eq__(self, b: str | int | bytes) -> bool:
```
Comment (yunzheng) [inline flow/record/fieldtypes/net/ip.py:27]:
```suggestion
    def __eq__(self, b: str | int | bytes | _IPAddress) -> bool:
```

File: flow/record/fieldtypes/net/ip.py [hunk 10]
```diff
@@ -1,56 +1,75 @@
...
 class ipnetwork(FieldType):
     val = None
     _type = "net.ipnetwork"
 
-    def __init__(self, addr):
+    def __init__(self, addr: str | int | bytes):
         self.val = ip_network(addr)
 
-    def __eq__(self, b):
+    def __eq__(self, b: str | int | bytes) -> bool:
```
Comment (yunzheng) [inline flow/record/fieldtypes/net/ip.py:62]:
```suggestion
    def __eq__(self, b: str | int | bytes | _IPNetwork) -> bool:
```

**Review by yunzheng — CHANGES_REQUESTED** [review]
some small suggestions and then it should be good to go!

**Review by yunzheng — APPROVED** [review]

**Pull request merged**